test/integration/udp_tunneling_integration_test.cc

```diff
@@ -712,6 +720,94 @@ TEST_P(UdpTunnelingIntegrationTest, ConnectionAttemptRetry) {
   test_server_->waitForGaugeEq("udp.foo.downstream_sess_active", 0);
 }
 
+TEST_P(UdpTunnelingIntegrationTest, PropagateResponseHeaders) {
+  const std::string access_log_filename =
+      TestEnvironment::temporaryPath(TestUtility::uniqueFilename());
+
+  const std::string session_access_log_config = fmt::format(R"EOF(
+  access_log:
+  - name: envoy.access_loggers.file
+    typed_config:
+      '@type': type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
+      path: {}
+      log_format:
+        text_format_source:
+          inline_string: "%FILTER_STATE(envoy.udp_proxy.propagate_response_headers:TYPED)%\n"
+)EOF",
+                                                            access_log_filename);
+
+  TestConfig config{"host.com",
+                    "target.com",
+                    1,
+                    30,
+                    false,
+                    "",
+                    BufferOptions{1, 30},
+                    absl::nullopt,
+                    session_access_log_config,
+                    true,
+                    false};
+  setup(config);
+
+  const std::string datagram = "hello";
+  establishConnection(datagram);
```

Comment:
Could you also one test case to cover invalid response?

Reply:
Added test.